frontend/src/scenes/data-warehouse/editor/draftsLogic.tsx

```diff
@@ -0,0 +1,158 @@
+        updateDraft: (draft: DataWarehouseSavedQueryDraft) => ({ draft }),
+        saveOrUpdateDraft: (query: HogQLQuery, viewId?: string, draftId?: string, editedHistoryId?: string) => ({
+            query,
+            viewId,
+            draftId,
+            editedHistoryId,
+        }),
+        deleteDraft: (draftId: string, successCallback?: () => void) => ({ draftId, successCallback }),
+        setDrafts: (drafts: DataWarehouseSavedQueryDraft[]) => ({ drafts }),
+        renameDraft: (draftId: string, name: string) => ({ draftId, name }),
+    }),
+
+    loaders(({ values }) => ({
+        draftsResponse: [
+            {} as PaginatedResponse<DataWarehouseSavedQueryDraft>,
+            {
+                loadDrafts: async () => {
+                    const drafts = await api.dataWarehouseSavedQueryDrafts.list()
+
+                    return drafts
+                },
+                loadMoreDrafts: async () => {
+                    if (values.draftsResponse.next) {
+                        const drafts = await api.get<PaginatedResponse<DataWarehouseSavedQueryDraft>>(
+                            values.draftsResponse.next
+                        )
+
+                        return {
+                            ...values.draftsResponse,
+                            results: [...values.draftsResponse.results, ...drafts.results],
+                            next: drafts.next,
+                        }
+                    }
+                    return values.draftsResponse
+                },
+            },
+        ],
+    })),
+    reducers({
+        drafts: [
+            [] as DataWarehouseSavedQueryDraft[],
+            {
+                setDrafts: (_, { drafts }) => drafts,
+            },
+        ],
+    }),
+    selectors({
+        hasMoreDrafts: [(s) => [s.draftsResponse], (draftsResponse) => draftsResponse.next !== null],
+    }),
+    listeners(({ values, actions }) => ({
+        loadDraftsSuccess: ({ draftsResponse }) => {
+            actions.setDrafts(draftsResponse.results)
+        },
+        loadMoreDraftsSuccess: ({ draftsResponse }) => {
+            actions.setDrafts(draftsResponse.results)
+        },
+        saveAsDraft: async ({ query, viewId, successCallback, editedHistoryId }) => {
+            const draft = await api.dataWarehouseSavedQueryDrafts.create({
+                query,
+                saved_query_id: viewId,
+                edited_history_id: editedHistoryId,
+            })
+            lemonToast.success('Draft saved')
+            successCallback && successCallback(draft)
+
+            const newDrafts = [...values.drafts, draft]
+            actions.setDrafts(newDrafts)
+        },
+        updateDraft: async ({ draft }) => {
+            await api.dataWarehouseSavedQueryDrafts.update(draft.id, draft)
+            lemonToast.success('Draft updated')
+            const newDrafts = values.drafts.map((d) => (d.id === draft.id ? draft : d))
+            actions.setDrafts(newDrafts)
```

Comment:
Great use of optimistic updates here, though we may wanna take the response of the `.update(...)` and use that to update `actions.setDrafts(newDrafts)` (e.g. things like timestamps would have updated and whatnot, its not a real issue now, but could create subtle bugs down the line)